lib/internal/modules/package_json_reader.js

```diff
@@ -1,30 +1,101 @@
 'use strict';
 
-const { SafeMap } = primordials;
+const {
+  JSONParse,
+  ObjectPrototypeHasOwnProperty,
+  SafeMap,
+} = primordials;
+const {
+  ERR_INVALID_PACKAGE_CONFIG,
+} = require('internal/errors').codes;
 const { internalModuleReadJSON } = internalBinding('fs');
-const { pathToFileURL } = require('url');
 const { toNamespacedPath } = require('path');
+const { kEmptyObject } = require('internal/util');
+const { fileURLToPath, pathToFileURL } = require('internal/url');
 
 const cache = new SafeMap();
 
 let manifest;
 
 /**
- *
  * @param {string} jsonPath
+ * @param {{
+ *   base?: string,
+ *   specifier: string,
+ *   isESM: boolean,
+ * }} options
+ * @returns {{
+ *   exists: boolean,
+ *   pjsonPath: string,
+ *   name?: string,
+ *   main?: string,
+ *   exports?: string | Record<string, unknown>,
+ *   imports?: string | Record<string, unknown>,
+ *   type: 'commonjs' | 'module' | 'none',
+ * }}
  */
-function read(jsonPath) {
+function read(jsonPath, { base, specifier, isESM } = kEmptyObject) {
   if (cache.has(jsonPath)) {
     return cache.get(jsonPath);
   }
 
-  const { 0: string, 1: containsKeys } = internalModuleReadJSON(
+  const string = internalModuleReadJSON(
     toNamespacedPath(jsonPath),
   );
-  const result = { string, containsKeys };
-  const { getOptionValue } = require('internal/options');
+  const result = {
+    __proto__: null,
+    exists: false,
+    pjsonPath: jsonPath,
+    main: undefined,
+    name: undefined,
+    type: 'none',
+    exports: undefined,
+    imports: undefined,
+  };
+
   if (string !== undefined) {
+    let parsed;
+    try {
+      parsed = JSONParse(string);
+    } catch (error) {
+      if (isESM) {
+        throw new ERR_INVALID_PACKAGE_CONFIG(
+          jsonPath,
+          (base ? `"${specifier}" from ` : '') + fileURLToPath(base || specifier),
+          error.message,
+        );
+      } else {
+        error.path = jsonPath;
+        error.message = 'Error parsing ' + jsonPath + ': ' + error.message;
```

Comment:
instead of mutating the error here, is it perhaps better to make a new one with error as the cause?

Reply:
You're right. I'm updating the pull request. Thanks for the review Jordan!

Reply:
Unfortunately, custom errors are disallowed by the linter.

```
/Users/yagiz/Developer/node/lib/internal/modules/package_json_reader.js
  73:19  error  Use an error exported by the internal/errors module  no-restricted-syntax
```

Reply:
ah, fair enough

Reply:
```suggestion
        // For backward compat, we modify the error returned by JSON.parse rather than creating a new one.
        // TODO(aduh95): make it throw ERR_INVALID_PACKAGE_CONFIG in a semver-major
        error.message = 'Error parsing ' + jsonPath + ': ' + error.message;
```

Reply:
Added the todo and the comment @aduh95